README.md
# concise-buffer
[![standard-readme compliant](https://img.shields.io/badge/readme%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
[![Semantic Versioning 2.0.0](https://img.shields.io/badge/semver-2.0.0-brightgreen?style=flat-square)](https://semver.org/spec/v2.0.0.html)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg?style=flat-square)](https://conventionalcommits.org)
[![License](https://img.shields.io/github/license/Anadian/concise-buffer)](https://github.com/Anadian/concise-buffer/blob/master/LICENSE)
[![npm](https://img.shields.io/npm/v/concise-buffer)](https://www.npmjs.com/package/concise-buffer)
[![CI](https://github.com/Anadian/concise-buffer/workflows/ci/badge.svg)](https://github.com/Anadian/concise-buffer/actions?query=workflow%3Aci)
[![Coverage Status](https://coveralls.io/repos/github/Anadian/concise-buffer/badge.svg?branch=master)](https://coveralls.io/github/Anadian/concise-buffer?branch=master)
![David](https://david-dm.org/Anadian/concise-buffer.svg)

> Concisely serialise the beginning and end of a Node buffer for logging.
# Table of Contents
- [Background](#Background)
- [Install](#Install)
- [Usage](#Usage)
- [API](#API)
- [Contributing](#Contributing)
- [License](#License)
# Background
I created this module simply because I was tired of Node Buffer's default serialisation producing massive garish noise in my logs. This module is incredibly simple: right now, it defines a single exported function, `getStringFromBuffer`, which turns an entire buffer into a nicely readable and consistent format `Buffer:[ <some_bytes> ... <some_bytes> ]`; how many bytes it includes can be specified and defaults to 16. If the number of bytes requested is greater than half the length of the buffer: it will simply serialise the whole buffer in hexadecimal.
# Install
To use it as a dependency in another package run:
```sh
npm install --save concise-buffer
```
To install it globally run:
```sh
npm install --global concise-buffer
```
# Usage
Example:
```js
const ConciseBuffer = require('concise-buffer');

const some_buffer = Buffer.from( [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 ] ); //In real use cases, the buffer will propbably be a lot longer than this but you get the idea.

var concise_string = ConciseBuffer.getStringFromBuffer( some_buffer, 4 ); //You can optionally specify how many bytes from the beginning and end of the buffer you want; the default is 16.

console.log( concise_string ); //Will print: `Buffer:[ 00010203 ... 0c0d0e0f ]`
```
# API
```js
const ConciseBuffer = require('concise-buffer');
```
See [API.md](API.md) for full API.
# Contributing
Changes are tracked in [CHANGELOG.md](CHANGELOG.md).
# License
MIT ©2021 Anadian

SEE LICENSE IN [LICENSE](LICENSE)

[![Creative Commons License](https://i.creativecommons.org/l/by-sa/4.0/88x31.png)](http://creativecommons.org/licenses/by-sa/4.0/)This project's documentation is licensed under a [Creative Commons Attribution-ShareAlike 4.0 International License](http://creativecommons.org/licenses/by-sa/4.0/).


CHANGES.md
2021-02-14 v0.0.0 First commit.


source/main.test.js
#!/usr/local/bin/node
'use strict';
/**
# [main.test.js](source/main.test.js)
> The test file for `main.js`.

Internal module name: `ConciseBufferTest`

Author: Anadian

Code license: MIT
```
	Copyright 2020 Anadian
	Permission is hereby granted, free of charge, to any person obtaining a copy of this 
software and associated documentation files (the "Software"), to deal in the Software 
without restriction, including without limitation the rights to use, copy, modify, 
merge, publish, distribute, sublicense, and/or sell copies of the Software, and to 
permit persons to whom the Software is furnished to do so, subject to the following 
conditions:
	The above copyright notice and this permission notice shall be included in all copies 
or substantial portions of the Software.
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Documentation License: [![Creative Commons License](https://i.creativecommons.org/l/by-sa/4.0/88x31.png)](http://creativecommons.org/licenses/by-sa/4.0/)
> The source-code comments and documentation are written in [GitHub Flavored Markdown](https://github.github.com/gfm/).

> The type notation used in this documentation is based off of the [Google Closure type system](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System).

> The status and feature lifecycle keywords used in this documentation are based off of my own standard [defined here](https://github.com/Anadian/FeatureLifeCycleStateStandard).
*/

//#Dependencies
	//##Internal
	const ConciseBuffer = require('./main.js');
	//##Standard
	const FileSystem = require('fs');
	//##External
	const AVA = require('ava');
//#Constants
//##Errors

//#Global Variables
/* istanbul ignore next */

//#Functions
AVA.before( function( t ){
	const random_buffer_1024 = FileSystem.readFileSync( 'test/random_1024.dat' );
	t.context.random_buffer_1024 = random_buffer_1024;
} );
AVA('setLogger:DisableLogging', function( t ){
	ConciseBuffer.setLogger( null );
	t.pass();
} );
AVA('getStringFromBuffer:Error:InvalidParam:input_buffer', function( t ){
	t.throws( ConciseBuffer.getStringFromBuffer.bind( null ), {
		instanceOf: TypeError,
		code: 'ERR_INVALID_ARG_TYPE'
	} );
	t.pass();
} );
AVA('getStringFromBuffer:Error:InvalidParam:number_of_bytes', function( t ){
	t.throws( ConciseBuffer.getStringFromBuffer.bind( t.context.random_buffer_1024, 'a string false' ), {
		instanceOf: TypeError,
		code: 'ERR_INVALID_ARG_TYPE'
	} );
	t.pass();
} );
AVA('getStringFromBuffer:Error:InvalidParam:options', function( t ){
	t.throws( ConciseBuffer.getStringFromBuffer.bind( t.context.random_buffer_1024, 8, 'a string false' ), {
		instanceOf: TypeError,
		code: 'ERR_INVALID_ARG_TYPE'
	} );
	t.pass();
} );
AVA('getStringFromBuffer:Success:DefaultLength', function( t ){
	t.is( ConciseBuffer.getStringFromBuffer( t.context.random_buffer_1024 ), 'Buffer:[ 8ae6d80ca5743723b87cf92945e24395 ... 8a9a710d39587d5d15b3756b35483ed8 ]' );
} );
AVA('getStringFromBuffer:Success:CustomLength', function( t ){
	t.is( ConciseBuffer.getStringFromBuffer( t.context.random_buffer_1024, 8 ), 'Buffer:[ 8ae6d80ca5743723 ... 15b3756b35483ed8 ]' );
} );
AVA('getStringFromBuffer:Success:SmallerBufferLengthThanNumberOfBytes', function( t ){
	const random_buffer_16 = t.context.random_buffer_1024.slice( 0, 16 );
	t.is( ConciseBuffer.getStringFromBuffer( random_buffer_16, 32 ), 'Buffer:[ 8ae6d80ca5743723b87cf92945e24395 ]' );
} );




//#Exports and Execution
if(require.main === module){
} else{
}


gulpfile.js
#!/usr/local/bin/node

const Gulp = require('gulp');
const GulpJSONEditor = require('gulp-json-editor');

exports['add-config-to-package-json'] = function addConfigToPackageJSON(){
	return Gulp.src('package.json').pipe(GulpJSONEditor( {
		scripts: {
			"test": "ava -v ./source/main.test.js",
			"lint": "eslint ./source/main.js",
			"generate-docs": "extract-documentation-comments -I source/main.js -O API.md",
			"update-config": "hjson -j ci/github-actions.hjson | json2yaml --preserve-key-order -o .github/workflows/ci.yml",
			"update-deps": "npm-check-updates -u",
			"release": "standard-version",
			"publish-release": "git push --follow-tags origin master && npm publish"
		},
		engines: {
			node: '>=12'
		},
		"eslintConfig": {
			"env": {
			"commonjs": true,
			"es6": true,
			"node": true
			},
			"extends": "eslint:recommended",
			"globals": {
			"Atomics": "readonly",
			"SharedArrayBuffer": "readonly"
			},
			"parserOptions": {
			"ecmaVersion": 11
			},
			"rules": {
			"no-unused-vars": [
				"warn"
			],
			"no-useless-escape": "off",
			"no-empty": "warn"
			}
		},
		"standard-version": {
			"types": [
			{
				"type": "build",
				"section": "Build",
				"hidden": false
			},
			{
				"type": "ci",
				"section": "Continuous Integration",
				"hidden": false
			},
			{
				"type": "chore",
				"section": "Chores",
				"hidden": false
			},
			{
				"type": "docs",
				"section": "Documentation",
				"hidden": false
			},
			{
				"type": "feat",
				"section": "Features",
				"hidden": false
			},
			{
				"type": "fix",
				"section": "Bug Fixes",
				"hidden": false
			},
			{
				"type": "perf",
				"section": "Performance",
				"hidden": false
			},
			{
				"type": "refactor",
				"section": "Refactoring",
				"hidden": false
			},
			{
				"type": "style",
				"section": "Code Style",
				"hidden": false
			},
			{
				"type": "test",
				"section": "Tests",
				"hidden": false
			}
			]
		}
	} ) ).pipe(Gulp.dest( '.' ));
};


source/main.js
#!/usr/local/bin/node
'use strict';
/**
# [concise-buffer.js](source/concise-buffer.js)
> Concisely serialise the beginning and end of a Node Buffer.

Internal module name: `Concise-Buffer`

Author: Anadian

Code license: MIT
```
	Copyright 2020 Anadian
	Permission is hereby granted, free of charge, to any person obtaining a copy of this 
software and associated documentation files (the "Software"), to deal in the Software 
without restriction, including without limitation the rights to use, copy, modify, 
merge, publish, distribute, sublicense, and/or sell copies of the Software, and to 
permit persons to whom the Software is furnished to do so, subject to the following 
conditions:
	The above copyright notice and this permission notice shall be included in all copies 
or substantial portions of the Software.
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Documentation License: [![Creative Commons License](https://i.creativecommons.org/l/by-sa/4.0/88x31.png)](http://creativecommons.org/licenses/by-sa/4.0/)
> The source-code comments and documentation are written in [GitHub Flavored Markdown](https://github.github.com/gfm/).

> The type notation used in this documentation is based off of the [Google Closure type system](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System).

> The status and feature lifecycle keywords used in this documentation are based off of my own standard [defined here](https://github.com/Anadian/FeatureLifeCycleStateStandard).
*/

//#Dependencies
	//##Internal
	//##Standard
	//##External
//#Constants
const FILENAME = 'concise-buffer.js';
const MODULE_NAME = 'Concise-Buffer';
var PACKAGE_JSON = {};
var PROCESS_NAME = '';
if(require.main === module)/* istanbul ignore next */{
	PROCESS_NAME = 'concise-buffer';
} else{
	PROCESS_NAME = process.argv0;
}
//##Errors

//#Global Variables
/* istanbul ignore next */
var Logger = { 
	log: () => {
		return null;
	}
};
//#Functions
/**
## Functions
*/
/**
### setLogger
> Allows this module's functions to log the given logger object.

Parametres:
| name | type | description |
| --- | --- | --- |
| logger | {?object} | The logger to be used for logging or `null` to disable logging. |

Throws:
| code | type | condition |
| --- | --- | --- |
| 'ERR_INVALID_ARG_TYPE' | {TypeError} | Thrown if `logger` is neither an object nor `null` |

History:
| version | change |
| --- | --- |
| 0.0.0 | Introduced |
*/
function setLogger( logger ){
	var return_error = null;
	//const FUNCTION_NAME = 'setLogger';
	//Variables
	//Parametre checks
	/* istanbul ignore else */
	if( typeof(logger) === 'object' ){
		/* istanbul ignore next */
		if( logger === null ){
			logger = { 
				log: () => {
					return null;
				}
			};
		}
	} else{
		return_error = new TypeError('Param "logger" is not an object.');
		return_error.code = 'ERR_INVALID_ARG_TYPE';
		throw return_error;
	}

	//Function
	Logger = logger;
	//Return
}
/**
### getStringFromBuffer
> Returns a string with the first few and last few bytes of the buffer.

Parametres:
| name | type | description |
| --- | --- | --- |
| input_buffer | {Buffer} | The Buffer to generate a concise string representation for.  |
| number_of_bytes | {number} | The number of bytes to show from the beginning and end of the buffer. If it's greater than half the length of the buffer, than the entire buffer will be serialised. \[default: \] |
| options | {?Object} | [Reserved] Additional run-time options. \[default: {}\] |

Returns:
| type | description |
| --- | --- |
| {string} | Returns a string of the format `Buffer:[ <first few hexadecimal bytes> ... <last few hexadecimal bytes> ]`  |

Throws:
| code | type | condition |
| --- | --- | --- |
| 'ERR_INVALID_ARG_TYPE' | {TypeError} | Thrown if a given argument isn't of the correct type. |

History:
| version | change |
| --- | --- |
| 0.0.1 | WIP |
*/
function getStringFromBuffer( input_buffer, number_of_bytes = 16, options = {},){
	var arguments_array = Array.from(arguments);
	var _return;
	var return_error;
	const FUNCTION_NAME = 'getStringFromBuffer';
	Logger.log({process: PROCESS_NAME, module: MODULE_NAME, file: FILENAME, function: FUNCTION_NAME, level: 'debug', message: `received: ${arguments_array}`});
	//Variables
	var beginning_string = '';
	var ending_string = '';
	//Parametre checks
	if( Buffer.isBuffer(input_buffer) === false ){
		return_error = new TypeError('Param "input_buffer" is not Buffer.');
		return_error.code = 'ERR_INVALID_ARG_TYPE';
		throw return_error;
	}
	if( typeof(number_of_bytes) !== 'number' ){
		return_error = new TypeError('Param "number_of_bytes" is not number.');
		return_error.code = 'ERR_INVALID_ARG_TYPE';
		throw return_error;
	}
	if( typeof(options) !== 'object' ){
		return_error = new TypeError('Param "options" is not ?Object.');
		return_error.code = 'ERR_INVALID_ARG_TYPE';
		throw return_error;
	}

	//Function
	if( (2 * number_of_bytes) < input_buffer.length ){
		beginning_string = input_buffer.toString( 'hex', 0, number_of_bytes );
		ending_string = input_buffer.toString( 'hex', (input_buffer.length - number_of_bytes) );
		_return = 'Buffer:[ '+beginning_string+' ... '+ending_string+' ]';
	} else{
		_return = 'Buffer:[ '+input_buffer.toString( 'hex' )+' ]';
	}

	//Return
	Logger.log({process: PROCESS_NAME, module: MODULE_NAME, file: FILENAME, function: FUNCTION_NAME, level: 'debug', message: `returned: ${_return}`});
	return _return;
}
//#Exports and Execution
if(require.main === module)/* istanbul ignore next */{
} else{
	exports.setLogger = setLogger;
	exports.getStringFromBuffer = getStringFromBuffer;
}
